### PR TITLE
Refactor ActiveRecord::Preloader associations

### PR DESCRIPTION
**Pull request by bogdan**
Removed the subclasses for Collection and Singular associations:
they are only different by one line of code which doesn't deserve a
hierarchy of 6 classes.
Now, the difference between singular and collection association looks like this:

``` ruby
association.target = reflection.collection? ? records : records.first
```

Introduced only `ThroughAssociation` and `DirectAssociation` inherited from
Association and implementing abstract `#run` method. The code is now well split between 2 subclasses. Also the inheritance is done in classic way without method override - only abstract method implementation.

r? @rafaelfranca

### CODE REVIEW

File: activerecord/lib/active_record/associations/preloader.rb [hunk 1]
```diff
@@ -181,22 +174,18 @@ def preloaded_records
...
-          when :has_many
-            reflection.options[:through] ? HasManyThrough : HasMany
-          when :has_one
-            reflection.options[:through] ? HasOneThrough : HasOne
-          when :belongs_to
-            BelongsTo
+          if reflection.options[:through]
+            ThroughAssociation
+          else
+            DirectAssociation
```
Comment (kamipo) [inline activerecord/lib/active_record/associations/preloader.rb]: Moving most methods to new class will hurt the blameability.
Prefer reusing `Preloader::Association` to touch lesser lines.

Comment (bogdan) [reply activerecord/lib/active_record/associations/preloader.rb]: Yes, I was thinking about that too, but things change... The blame is still traceable manually via tracking particular line of code via `git log -p`. When I backtrace rails changes, it is almost always the case. The code moved from `Association` to `DirectAssociation` is easy to read and it didn't change for a really long period of time. So decided to sacrifice blame for ideal OOD.

Comment (matthewd) [reply activerecord/lib/active_record/associations/preloader.rb]: If it's moving mostly in unmodified hunks, I believe blame can follow it across files anyway.

(I haven't looked at the actual diff, so have no opinion on that beyond "+117 −195" sounding good.)

Comment (kamipo) [reply activerecord/lib/active_record/associations/preloader.rb]: At least we can remove preloader classes without touching existing code.
https://github.com/rails/rails/compare/master...kamipo:remove_useless_preloader_classes

Comment (rafaelfranca) [reply activerecord/lib/active_record/associations/preloader.rb]: I'm more inclined to kamipo's refactoring.

Comment (bogdan) [reply activerecord/lib/active_record/associations/preloader.rb]: That is fine. You may accept that one.

Comment (rafaelfranca) [reply activerecord/lib/active_record/associations/preloader.rb]: @kamipo mind to take care of it?

Comment (kamipo) [reply activerecord/lib/active_record/associations/preloader.rb]: Yeah, I will do that.

**Pull request closed without merging**